data_model/src/asset.rs

```diff
@@ -150,21 +150,31 @@ pub enum AssetValueType {
     Store,
 }
 
-impl FromStr for AssetValueType {
-    type Err = &'static str;
-
-    fn from_str(value_type: &str) -> Result<Self, Self::Err> {
-        // TODO: Could be implemented with some macro
-        match value_type {
-            "Quantity" => Ok(AssetValueType::Quantity),
-            "BigQuantity" => Ok(AssetValueType::BigQuantity),
-            "Fixed" => Ok(AssetValueType::Fixed),
-            "Store" => Ok(AssetValueType::Store),
-            _ => Err("Unknown variant"),
+/// A declarative macro that implements `FromStr` for a given
+/// C like enumeration. The macro is invoked like follows:
+/// `easy_from_str_impl! { NameOfEnum, EnumVariation1, EnumVariation2, ... }`
+macro_rules! easy_from_str_impl {
+    (eval_to $cmp:expr, $enum_type:ty, $enum_value:tt) => {
+        if $cmp == stringify!($enum_value) {
```

Comment:
it's better to use match statement because it will most likely be compiled into a jump table. With `if-else` statements I wouldn't be so sure of it. If you're having trouble with implementing it as a declarative macro, I think you can find examples of generating match statements with declarative macros in the codebase

Reply:
also, I would be more in favor of using procedural macro here which would guarantee exhaustive match if new variant is added. I don't think you can make such a guarantee with a declarative macro

I don't think we've got much from this macro in it's current form

Reply:
What Sam can do here is use this as a reference implementation for an Iroha macro procedural macro. He could also do that as a separate ticket.

Reply:
> it's better to use match statement because it will most likely be compiled into a jump table. With `if-else` statements I wouldn't be so sure of it. If you're having trouble with implementing it as a declarative macro, I think you can find examples of generating match statements with declarative macros in the codebase

It is not possible to produce a jump table for string comparison. Maybe if you were comparing string literals only and knew that the same string would have the same pointer into the .text block in the binary you could. But &str has no such guarantee and so the compiler output will be the same for match and multiple if's.

Reply:
I've created the seperate ticket #2056 for the tracking of the better macro.

Reply:
> It is not possible to produce a jump table for string comparison. Maybe if you were comparing string literals only and knew that the same string would have the same pointer into the .text block in the binary you could. But &str has no such guarantee and so the compiler output will be the same for match and multiple if's.

ah well, perfect hashing approach would make it possible to produce jump table even for strings. Though it may be unlikely compiler would make such an "optimization"

Reply:
it doesn't even have to be a perfect hash

Reply:
While this conversation is resolved, I would have to concede, that a procedural macro can use a hash map (or literal interning) to make the comparison a lightweight jump.